src/amd/compiler/aco_util.h

```diff
@@ -0,0 +1,235 @@
+	    : data{ data } , length{ length } {}
+
+	/*!	\brief						Returns an iterator to the begin of the span
+	*	\return						data
+	*/
+	iterator begin() noexcept {
+		return data;
+	}
+
+	/*!	\brief						Returns a const_iterator to the begin of the span
+	*	\return						data
+	*/
+	const_iterator begin() const noexcept {
+		return data;
+	}
+
+	/*!	\brief						Returns an iterator to the end of the span
+	*	\return						data + length
+	*/
+	iterator end() noexcept {
+		return std::next(data, length);
+	}
+
+	/*!	\brief						Returns a const_iterator to the end of the span
+	*	\return						data + length
+	*/
+	const_iterator end() const noexcept {
+		return std::next(data, length);
+	}
+
+	/*!	\brief						Returns a const_iterator to the begin of the span
+	*	\return						data
+	*/
+	const_iterator cbegin() const noexcept {
+		return data;
+	}
+
+	/*!	\brief						Returns a const_iterator to the end of the span
+	*	\return						data + length
+	*/
+	const_iterator cend() const noexcept {
+		return std::next(data, length);
+	}
+
+	/*!	\brief						Returns a reverse_iterator to the end of the span
+	*	\return						reverse_iterator(end())
+	*/
+	reverse_iterator rbegin() noexcept {
+		return reverse_iterator(end());
+	}
+
+	/*!	\brief						Returns a const_reverse_iterator to the end of the span
+	*	\return						reverse_iterator(end())
+	*/
+	const_reverse_iterator rbegin() const noexcept {
+		return const_reverse_iterator(end());
+	}
+
+	/*!	\brief						Returns a reverse_iterator to the begin of the span
+	*	\return						reverse_iterator(begin())
+	*/
+	reverse_iterator rend() noexcept {
+		return reverse_iterator(begin());
+	}
+
+	/*!	\brief						Returns a const_reverse_iterator to the begin of the span
+	*	\return						reverse_iterator(begin())
+	*/
+	const_reverse_iterator rend() const noexcept {
+		return const_reverse_iterator(begin());
+	}
+
+	/*!	\brief						Returns a const_reverse_iterator to the end of the span
+	*	\return						rbegin()
+	*/
+	const_reverse_iterator crbegin() const noexcept {
+		return const_reverse_iterator(cend());
+	}
+
+	/*!	\brief						Returns a const_reverse_iterator to the begin of the span
+	*	\return						rend()
+	*/
+	const_reverse_iterator crend() const noexcept {
+		return const_reverse_iterator(cbegin());
+	}
+
+	/*!	\brief						Unchecked access operator
+	*	\param[in]	index			Index of the element we want to access
+	*	\return						*(std::next(data, index))
+	*/
+	reference operator[](const size_type index) noexcept(false) {
+		return *(std::next(data, index));
+	}
+
+	/*!	\brief						Unchecked const access operator
+	*	\param[in]	index			Index of the element we want to access
+	*	\return						*(std::next(data, index))
+	*/
+	const_reference operator[](const size_type index) const noexcept(false) {
+		return *(std::next(data, index));
+	}
+
+	/*!	\brief						Returns a reference to the last element of the span
+	*	\return						*end()
+	*/
+	reference back() noexcept(false) {
```

Comment:
Why `noexcept(false)`? This (and others) look like `noexcept(true)`.

Reply:
A `span` is non-owning. Consequently there is no guarantee that the underlying data is valid. Also it is valid to create an empty span as well as clear it

Reply:
Actually you are right. While it would access invalid memory it would never throw an exception. I changed the front/back/operator[] methods accordingly